src/vs/workbench/contrib/tasks/test/browser/taskTerminalStatus.test.ts

```diff
@@ -0,0 +1,74 @@
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { strictEqual } from 'assert';
+import { Emitter, Event } from 'vs/base/common/event';
+import { TestConfigurationService } from 'vs/platform/configuration/test/common/testConfigurationService';
+import { TestInstantiationService } from 'vs/platform/instantiation/test/common/instantiationServiceMock';
+import { ACTIVE_TASK_STATUS, FAILED_TASK_STATUS, SUCCEEDED_TASK_STATUS, TaskTerminalStatus } from 'vs/workbench/contrib/tasks/browser/taskTerminalStatus';
+import { AbstractProblemCollector } from 'vs/workbench/contrib/tasks/common/problemCollectors';
+import { CommonTask, TaskEvent, TaskEventKind, TaskRunType } from 'vs/workbench/contrib/tasks/common/tasks';
+import { ITaskService, Task } from 'vs/workbench/contrib/tasks/common/taskService';
+import { ITerminalInstance } from 'vs/workbench/contrib/terminal/browser/terminal';
+import { TerminalStatusList } from 'vs/workbench/contrib/terminal/browser/terminalStatusList';
+
+class TestTaskService implements Partial<ITaskService> {
+	private readonly _onDidStateChange: Emitter<TaskEvent> = new Emitter();
+	public get onDidStateChange(): Event<TaskEvent> {
+		return this._onDidStateChange.event;
+	}
+	public triggerStateChange(event: TaskEvent): void {
+		this._onDidStateChange.fire(event);
+	}
+}
+
+class TestTerminal implements Partial<ITerminalInstance> {
+	statusList: TerminalStatusList = new TerminalStatusList(new TestConfigurationService());
+}
+
+class TestTask extends CommonTask {
+	protected getFolderId(): string | undefined {
+		throw new Error('Method not implemented.');
+	}
+	protected fromObject(object: any): Task {
+		throw new Error('Method not implemented.');
+	}
+}
+
+class TestProblemCollector implements Partial<AbstractProblemCollector> {
+
+}
+
+const instantiationService = new TestInstantiationService();
+suite('Task Terminal Status', () => {
+	let taskService: TestTaskService;
+	let taskTerminalStatus: TaskTerminalStatus;
+	let testTerminal: ITerminalInstance;
+	let testTask: Task;
+	let problemCollector: AbstractProblemCollector;
+	setup(() => {
+		taskService = new TestTaskService();
+		taskTerminalStatus = instantiationService.createInstance(TaskTerminalStatus, taskService);
+		testTerminal = instantiationService.createInstance(TestTerminal);
+		testTask = instantiationService.createInstance(TestTask);
+		problemCollector = instantiationService.createInstance(TestProblemCollector);
+	});
+	test('Should add failed status when there is an exit code on task end', async () => {
+		taskTerminalStatus.addTerminal(testTask, testTerminal, problemCollector);
+		taskService.triggerStateChange({ kind: TaskEventKind.End, exitCode: 1 });
+		setTimeout(() => strictEqual(testTerminal.statusList.statuses, [FAILED_TASK_STATUS]), 200);
```

Comment:
these `setTimeout` calls are not what we want, but without them it seems the state of the terminal status list doesn't have time to update - it's a little strange though bc it is only severity that doesn't get updated
```
 1) Task Terminal Status
       Task Terminal Status
         Should add failed status when there is an exit code on task end:

      
      + expected - actual

       [
         {
      +    "icon": {
      +      "definition": {
      +        "fontCharacter": "\\ea87"
      +      }
      +      "description": "[undefined]"
      +      "id": "error"
      +    }
           "id": "task_terminal_status"
      -    "severity": 1
      +    "severity": 3
      +    "tooltip": "Task has errors"
         }
       ]
```

Reply:
I also checked in OSS and can't repro a case where this is not updated immediately

thoughts @Tyriar

Reply:
it would help if I could log in tests, but I don't see any console.logs in the test output

Reply:
oh looks like strictEqual is part of the problem

Reply:
working on it